File: javascript/easy/trimABinarySearchTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} L
 * @param {number} R
 * @return {TreeNode}
 */
var trimBST = function(root, L, R) {
    if(root === null) return null;
    root = validRoot(root, L, R);
    if(root !== null) {
        root.left = trimBST(root.left, L, R);
        root.right = trimBST(root.right, L, R);
    }
    return root;
};

var validRoot = function(root, L, R) {
    if(root === null) return null;
    if(root.val >= L && root.val <= R) {
        return root;
    } else if(root.val < L) {
        return validRoot(root.right, L, R);
    } else {
        return validRoot(root.left, L, R);
    }
};

File: ruby/easy/trim_a_binary_search_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} l
# @param {Integer} r
# @return {TreeNode}
def trim_bst(root, l, r)
    return nil unless root
    while(root && !(root.val >= l && root.val <= r))
        if(root.val < l)
            root = root.right
        else
            root = root.left
        end
    end
    if(root)
        root.left = trim_bst(root.left, l, r)
        root.right = trim_bst(root.right, l, r)
    end
    return root
end

